src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java

```diff
@@ -250,121 +240,65 @@ public void putDataPoint(String metricName,
-                    bs.setString(0, metricName);
-                    bs.setString(1, split);
-                    bs.setString(2, v);
-                    bs.setBytes(3, serializedKey);
-                    bs.setInt(4, rowKeyTtl);
-
-                    m_session.executeAsync(bs);
+            boolean writeRowKeyReverseLookups = false;
+            final boolean rowKeyKnown = rowKeyCache.isKnown(serializedKey);
+            if (!rowKeyKnown) {
+                writeRowKeyReverseLookups = true;
+                rowKeyCache.put(serializedKey);
+
+                int count = ++keyInsertCount;
+                if (insertCountByMetric.containsKey(metricName)) {
+                    insertCountByMetric.put(metricName, insertCountByMetric.get(metricName) + 1);
+                } else {
+                    insertCountByMetric.put(metricName, 1);
                 }
 
-                for (RowKeyListener rowKeyListener : m_rowKeyListeners) {
-                    rowKeyListener.addRowKey(metricName, rowKey, rowKeyTtl);
+                if ((writeTime - keyInsertTimer) > 30_000) {
+                    keyInsertTimer = writeTime;
+                    keyInsertCount = 0;
+                    countsByMetricName = new ArrayList<>(insertCountByMetric.entrySet());
+                    insertCountByMetric.clear();
+                    logger.warn("RowKeys inserted: count={}", count);
                 }
+            }
 
-                if (countsByMetricName != null) {
-                    Collections.sort(countsByMetricName, (x,y) -> y.getValue() - x.getValue());
-                    logger.warn("Keys inserted: {}", countsByMetricName.subList(0, Math.min(10, countsByMetricName.size())));
-                }
+            if (writeRowKeyReverseLookups) {
+                storeRowKeyReverseLookups(metricName, serializedKey, rowKeyTtl, tags, countsByMetricName);
             }
 
             //Write metric name if not in cache
-            Boolean isCachedName  = m_metricNameCache.getIfPresent(metricName);
-            if (isCachedName == null) {
-                m_metricNameCache.put(metricName, CACHE_BOOLEAN);
+            if (!metricNameCache.isKnown(metricName)) {
                 if (metricName.length() == 0) {
-                    logger.warn(
-                            "Attempted to add empty metric name to string index. Row looks like: " + dataPoint
-                    );
+                    logger.warn("Attempted to add empty metric name to string index. Row looks like: {}", dataPoint);
                 }
-                BoundStatement bs = new BoundStatement(m_psInsertString);
-                bs.setBytes(0, ByteBuffer.wrap(ROW_KEY_METRIC_NAMES.getBytes(UTF_8)));
-                bs.setString(1, metricName);
-                m_session.executeAsync(bs);
+                metricNameCache.put(metricName);
+                storeStringIndex(metricName, m_psInsertString, METRIC_NAME_BYTE_BUFFER);
```

Comment:
I would rather store to the database first and then update the cache. In case of Cassandra failure we will have inconsistent state in storage: cache will have something while database will not. The same goes to `tagNameCache` below

Reply:
Sounds legit. Will change to do that. Thanks